### PR TITLE
Clean up windows desktop access error logs on expected disconnects

### PR DESCRIPTION
**Pull request by ibeckermayer**
Refactors the error handling code in windows desktop sessions such that server and client initiated disconnects no longer show errors in the logs. 

This addresses https://github.com/gravitational/teleport/issues/14451, although not exactly in the way that it specifies. Client initiated disconnects are still not sending a [MCS Disconnect Provider Ultimatum PDU](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpbcgr/27915739-8f77-487e-9927-55008af7fd68) before closing the connection.

Note to self: do not merge until
- [x] https://github.com/gravitational/rdp-rs/pull/26 has been merged
- [x] and git ref in Cargo.toml changed to the new head

### CODE REVIEW

File: lib/web/desktop.go
Comment (codingllama) [inline lib/web/desktop.go:395]:
```suggestion
// handleProxyWebsocketConnErr handles the error returned by proxyWebsocketConn.
```

Handles the error how? Is it worth elaborating a bit?

File: lib/web/desktop.go
Comment (codingllama) [inline lib/web/desktop.go:396]: Rename `err` to something else, so it stands out in relation to local errors? `websocketErr` maybe?

File: lib/web/desktop.go
Comment (codingllama) [inline lib/web/desktop.go:415]:
```suggestion
			log.WithError(err).Warning("Error proxying a desktop protocol websocket to windows_desktop_service")
```

File: lib/web/desktop.go
Comment (codingllama) [inline lib/web/desktop.go:418]:
```suggestion
		log.Error("Programming error: received unexpected error type from proxyWebsocketConn")
```

File: lib/web/desktop.go
Comment (codingllama) [inline lib/web/desktop.go:401]: Handle `!ok` in the if and return immediately, so we remove the nesting?

https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow

File: lib/web/desktop.go
Comment (codingllama) [inline lib/web/desktop.go:408]: Should we stop here?

File: lib/web/desktop.go
Comment (codingllama) [inline lib/web/desktop.go:401]: Should we type switch and loop over `errors.Unwrap`, for added resilience to other types of wrapping?

Eg, something like this:

```go
errs = []error{websocketErr}
for len(errs) > 0 {
	err := errs[0] // pop first error
	errs = errs[1:]

	switch err := err.(type) {
	case trace.Aggregate:
		errs = append(errs, err.Errors())
	case *websocket.CloseError:
		// Check codes, log, etc.
		return
	default:
		if wrapped := errors.Unwrap(err); wrapped != nil {
			errs = append(errs, wrapped)
		}
	}
}
```

File: lib/srv/desktop/rdp/rdpclient/src/lib.rs
Comment (codingllama) [inline lib/srv/desktop/rdp/rdpclient/src/lib.rs:1137]: Could we have unambiguous codes for client and server disconnects? As soon as we add a message to client disconnects this scheme breaks.

File: lib/srv/desktop/rdp/rdpclient/client.go
Comment (ibeckermayer) [reply lib/srv/desktop/rdp/rdpclient/client.go:284]: Sometimes the message here will be something benign, like "The disconnection was initiated by an administrative tool on the server in another session.", whereas other times it can be an error, like "An internal error has occurred in the Terminal Services licensing component.". It seems like overkill to distinguish between [all the possibilities](https://github.com/gravitational/rdp-rs/blob/389eaf7bb8f207c1e4f98ed3f2766816480ee912/src/core/global.rs#L281-L317) here, so I just went with a warning.

The api for `read_rdp_output` is admittedly a bit weird, its explained in the docstring. Open to suggestions for how to make it better

https://github.com/gravitational/teleport/blob/b7355f8bec35f8662239c23b9bf389ef7b0ad192/lib/srv/desktop/rdp/rdpclient/src/lib.rs#L1134-L1150

File: lib/srv/desktop/rdp/rdpclient/src/lib.rs
Comment (ibeckermayer) [reply lib/srv/desktop/rdp/rdpclient/src/lib.rs:1137]: Yeah, I'll add a `DisconnectCode`

File: lib/srv/desktop/rdp/rdpclient/client.go
Comment (codingllama) [reply lib/srv/desktop/rdp/rdpclient/client.go:284]: The disconnect code you talked about helps. A small comment here could help too, something saying that sometimes the message might be an error like the one you described above, hence we go with a warning.

Comment (ibeckermayer) [reply lib/srv/desktop/rdp/rdpclient/client.go]: Scratch all that, I added a function to distinguish between errors and just messages in rdp-rs and updated the handling and logging in 2598c7f3626e74013dd2b5fed943853db8691bd3

File: lib/srv/desktop/rdp/rdpclient/src/lib.rs
Comment (ibeckermayer) [reply lib/srv/desktop/rdp/rdpclient/src/lib.rs]: 2598c7f3626e74013dd2b5fed943853db8691bd3

File: lib/srv/desktop/rdp/rdpclient/client.go
Comment (ibeckermayer) [reply lib/srv/desktop/rdp/rdpclient/client.go]: 2598c7f3626e74013dd2b5fed943853db8691bd3

File: lib/srv/desktop/rdp/rdpclient/client.go
Comment (zmb3) [inline lib/srv/desktop/rdp/rdpclient/client.go:292]: In this case, since the disconnect reason never made it to the user, we should probably log it.

File: lib/srv/desktop/rdp/rdpclient/client.go
Comment (zmb3) [inline lib/srv/desktop/rdp/rdpclient/client.go:311]: If you set log to `Infof` or `Errorf` you could avoid the `Sprintf` call here.

File: lib/srv/desktop/rdp/rdpclient/src/lib.rs
Comment (zmb3) [inline lib/srv/desktop/rdp/rdpclient/src/lib.rs:1410]:
```suggestion
                debug!("TCP socket was not connected");
```

File: lib/web/desktop.go
Comment (zmb3) [inline lib/web/desktop.go:406]: Should we update the UI to actually send a close when the user hits disconnect in the menu?

File: lib/web/desktop.go
Comment (ibeckermayer) [reply lib/web/desktop.go]: d9d1d6f87852a508b34da1da7665c013ff785feb

File: lib/web/desktop.go
Comment (ibeckermayer) [reply lib/web/desktop.go]: d9d1d6f87852a508b34da1da7665c013ff785feb

File: lib/web/desktop.go
Comment (ibeckermayer) [reply lib/web/desktop.go]: d9d1d6f87852a508b34da1da7665c013ff785feb

File: lib/web/desktop.go
Comment (ibeckermayer) [reply lib/web/desktop.go]: d9d1d6f87852a508b34da1da7665c013ff785feb

File: lib/web/desktop.go
Comment (ibeckermayer) [reply lib/web/desktop.go]: d9d1d6f87852a508b34da1da7665c013ff785feb

File: lib/srv/desktop/rdp/rdpclient/client.go
Comment (ibeckermayer) [reply lib/srv/desktop/rdp/rdpclient/client.go:292]: It will get logged in the section below

File: lib/web/desktop.go
Comment (ibeckermayer) [reply lib/web/desktop.go]: 846afbe85ad0fda356857743a2346a1ddb534877 + https://github.com/gravitational/webapps/pull/1442

File: lib/srv/desktop/rdp/rdpclient/client.go
Comment (codingllama) [inline lib/srv/desktop/rdp/rdpclient/client.go:281]: nit:

```suggestion
		// Copy the returned message and free the C memory.
```

File: lib/srv/desktop/rdp/rdpclient/Cargo.toml
Comment (codingllama) [inline lib/srv/desktop/rdp/rdpclient/Cargo.toml:23]: Temporary hack?

File: lib/srv/desktop/rdp/rdpclient/src/lib.rs
Comment (codingllama) [inline lib/srv/desktop/rdp/rdpclient/src/lib.rs:1434]: Much clearer, thanks.

File: lib/srv/desktop/rdp/rdpclient/Cargo.toml
Comment (codingllama) [reply lib/srv/desktop/rdp/rdpclient/Cargo.toml:23]: https://github.com/gravitational/rdp-rs/pull/26, right? I'll resolve.

**Review by codingllama — APPROVED** [review]
Leaving approved so you are not blocked after the prereqs land.

**Review by zmb3 — APPROVED** [review]

**Pull request merged**
